java-manta-client/src/main/java/com/joyent/manta/client/multipart/AbstractMultipartManager.java

```diff
@@ -0,0 +1,199 @@
+    }
+
+    @Override
+    public PART uploadPart(final UPLOAD upload,
+                           final int partNumber,
+                           final File file) throws IOException {
+        validatePartNumber(partNumber);
+        Validate.notNull(file, "File must not be null");
+
+
+        if (!file.exists()) {
+            String msg = String.format("File doesn't exist: %s",
+                    file.getPath());
+            throw new FileNotFoundException(msg);
+        }
+
+        if (!file.canRead()) {
+            String msg = String.format("Can't access file for read: %s",
+                    file.getPath());
+            throw new IOException(msg);
+        }
+
+        HttpEntity entity = new FileEntity(file, ContentType.APPLICATION_OCTET_STREAM);
+        return uploadPart(upload, partNumber, entity);
+    }
+
+    @Override
+    public PART uploadPart(final UPLOAD upload,
+                           final int partNumber,
+                           final InputStream inputStream)
+            throws IOException {
+        Validate.notNull("InputStream must not be null");
+
+        if (inputStream.getClass().equals(FileInputStream.class)) {
+            final FileInputStream fin = (FileInputStream)inputStream;
+            final long contentLength = fin.getChannel().size();
+            return uploadPart(upload, partNumber, contentLength, inputStream);
+        }
+
+        HttpEntity entity = new MantaInputStreamEntity(inputStream,
+                ContentType.APPLICATION_OCTET_STREAM);
+
+        return uploadPart(upload, partNumber, entity);
+    }
+
+    @Override
+    public PART uploadPart(final UPLOAD upload,
+                           final int partNumber,
+                           final long contentLength,
+                           final InputStream inputStream) throws IOException {
+        Validate.notNull("InputStream must not be null");
+
+        HttpEntity entity;
+
+        if (contentLength > -1) {
+            entity = new MantaInputStreamEntity(inputStream, contentLength,
+                    ContentType.APPLICATION_OCTET_STREAM);
+        } else {
+            entity = new MantaInputStreamEntity(inputStream,
+                    ContentType.APPLICATION_OCTET_STREAM);
+        }
+
+        return uploadPart(upload, partNumber, entity);
+    }
+
+    /**
+     * Uploads a single part of a multipart upload.
+     *
+     * @param upload multipart upload object
+     * @param partNumber part number to identify relative location in final file
+     * @param entity Apache HTTP Client entity instance
+     * @return multipart single part object
+     * @throws IOException thrown if there is a problem connecting to Manta
+     */
+    abstract PART uploadPart(UPLOAD upload,
+                             int partNumber,
+                             HttpEntity entity) throws IOException;
+
+    /**
+     * Uses reflection to read a field from a MantaClient instance. This method
+     * is used to access fields
```

Comment:
I think this comment should explain why this approach is preferred over making the `MantaClient` instance available to subclasses.